controllers/apidefinition_controller.go

```diff
@@ -172,49 +197,71 @@ func (r *ApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 	return ctrl.Result{RequeueAfter: queueA}, err
 }
 
+func uploadCert(ctx context.Context, orgID string, pemKeyBytes, pemCrtBytes []byte) (tykCertID string, err error) {
+	tykCertID = orgID + cert.CalculateFingerPrint(pemCrtBytes)
+	exists := klient.Universal.Certificate().Exists(ctx, tykCertID)
+
+	if !exists {
+		// upload the certificate
+		tykCertID, err = klient.Universal.Certificate().Upload(ctx, pemKeyBytes, pemCrtBytes)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	return tykCertID, nil
+}
+
 func (r *ApiDefinitionReconciler) checkSecretAndUpload(
 	ctx context.Context,
 	certName string,
-	namespacedName types.NamespacedName,
+	ns string,
 	log logr.Logger,
 	env *environmet.Env,
+	field string,
 ) (string, error) {
 	secret := v1.Secret{}
 
-	err := r.Get(ctx, types.NamespacedName{Name: certName, Namespace: namespacedName.Namespace}, &secret)
+	err := r.Get(ctx, types.NamespacedName{Name: certName, Namespace: ns}, &secret)
 	if err != nil {
 		log.Error(err, "requeueing because secret not found")
 		return "", err
 	}
 
-	pemCrtBytes, ok := secret.Data["tls.crt"]
-	if !ok {
-		err = fmt.Errorf("%s", "requeueing because cert not found in secret")
-		log.Error(err, "requeueing because cert not found in secret")
+	// 'field' holds Secret object's data field that we are going to check.
+	// If kubernetes.io/tls type Secret is used, there is no need to provide 'field' since kubernetes.io/tls types have
+	// already defined fields called 'tls.crt' and 'tls.key'. However, if you provide any other type of secrets such
+	// as Opaque (as we do in Public Key Certificate pinning), you need to provide field that controller should check,
+	// through 'public_key_secret_field'.
+	if secret.Type == TLSSecretType {
+		pemCrtBytes, ok := secret.Data["tls.crt"]
+		if !ok {
+			err = fmt.Errorf("%s", "requeueing because cert not found in secret")
+			log.Error(err, "requeueing because cert not found in secret")
 
-		return "", err
+			return "", err
+		}
+
+		pemKeyBytes, ok := secret.Data["tls.key"]
+		if !ok {
+			err = fmt.Errorf("%s", "requeueing because key not found in secret")
+			log.Error(err, "requeueing because key not found in secret")
+
+			return "", err
+		}
+
+		return uploadCert(ctx, env.Org, pemKeyBytes, pemCrtBytes)
 	}
 
-	pemKeyBytes, ok := secret.Data["tls.key"]
+	pubKey, ok := secret.Data[field]
 	if !ok {
-		err = fmt.Errorf("%s", "requeueing because key not found in secret")
+		err = fmt.Errorf("requeueing because %s not found in secret", field)
```

Comment:
I think we should just log error here. Requeueing or not should be left to the function calling this function.

Reply:
This part becomes deprecated once we remove Opaque support. Therefore, I kept the logic here as it is